### PR TITLE
Add async API (2)

### PR DESCRIPTION
**Pull request by katcharov**
JAVA-5082

This is https://github.com/mongodb/mongo-java-driver/pull/1234, but with demo refactoring removed (that is, this is just the async API, with no other changes).

The first commit, "Add async functions", was reviewed as part of https://github.com/mongodb/mongo-java-driver/pull/1134 , but can be further reviewed. Other commits are all new non-reviewed code.

### CODE REVIEW

**Review by jyemin — CHANGES_REQUESTED** [review]
A few comments on the second commit

File: driver-core/src/main/com/mongodb/internal/async/AsyncSupplier.java [hunk 1]
```diff
@@ -108,13 +107,13 @@ default AsyncRunnable thenConsume(final AsyncConsumer<T> consumer) {
 
     /**
      * @param errorCheck A check, comparable to a catch-if/otherwise-rethrow
-     * @param supplier   The branch to execute if the error matches
+     * @param errorFunction   The branch to execute if the error matches
```
[the remark this answers is not in the document]
Comment (katcharov) [reply driver-core/src/main/com/mongodb/internal/async/AsyncSupplier.java:110]: There are both errorCheck and errorFunction, I think this is correct?

File: driver-core/src/main/com/mongodb/internal/async/SingleResultCallback.java [hunk 2]
```diff
@@ -34,4 +36,32 @@ public interface SingleResultCallback<T> {
...
+
+    default void complete(final SingleResultCallback<Void> callback) {
+        // takes a void callback (itself) to help ensure that this method
+        // is not accidentally used when "complete(T)" should have been used
+        // instead, since results are not marked nullable.
+        Assertions.assertTrue(callback == this);
+        this.onResult(null, null);
+    }
+
+    default void complete(final T result) {
```
[the remark this answers is not in the document]
Comment (katcharov) [reply driver-core/src/main/com/mongodb/internal/async/SingleResultCallback.java:64]: It seems cleaner to `throw`, since there is no chance that an ensuing `return` will be forgotten.

(This method would not be needed, except in some cases, there can be a non-async branch that needs to complete using the callback.)

File: driver-core/src/main/com/mongodb/internal/async/AsyncSupplier.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jyemin) [reply driver-core/src/main/com/mongodb/internal/async/AsyncSupplier.java:110]: I think I misread the diff just looking at the last commit.  Sorry for the noise.

File: driver-core/src/main/com/mongodb/internal/async/SingleResultCallback.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jyemin) [reply driver-core/src/main/com/mongodb/internal/async/SingleResultCallback.java:64]: I was thinking that code like this in `AsyncRunnable`:

```
            } catch (Throwable t) {
                callback.onResult(null, t);
                return;
            }
```

would be cleaner as

```
            } catch (Throwable t) {
                callback.completeExceptionally(t);
                return;
            }
```

**Review by jyemin — COMMENTED** [review]
Resolve one thread and commented back on another.

File: driver-core/src/main/com/mongodb/internal/async/SingleResultCallback.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (katcharov) [reply driver-core/src/main/com/mongodb/internal/async/SingleResultCallback.java:64]: This is true, but I think having it a less clean internally is better than exposing that second option externally, when it seems that in all cases we should prefer to write:
```
            } catch (Throwable t) {
                throw t;
            }
```
We can't do this internally in `AsyncRunnable`, because that is the place where thrown exceptions are caught and handled. And we can't have `completeExceptionally` was a private method on an interface, to prevent it being exposed. (Perhaps there is an external case where we cannot use `throw`, or where this is not cleaner?)

[the remark this answers is not in the document]
Comment (jyemin) [reply driver-core/src/main/com/mongodb/internal/async/SingleResultCallback.java:64]: I see your point for code that has been written using the new framework, but it still seems like a good thing to add for SingleResultCallback-based code where we're not quite ready to re-write using the new framework.  Everywhere that we currently have:

```
(result, t) -> { 
           if (t != null) {
                callback.onResult(null, t);
            } else {
               ....
               callback.onResult(newVal, null);
            }
```

would be clearer as

```
(result, t) -> { 
           if (t != null) {
                callback.completeExceptionally(t);
            } else {
               ....
               callback.complete(newVal);
            }
```

If we had been able to use default methods back when `SingleResultCallack` was first introduced (I think we still supported Java 6 back then?) we probably would have added then straightaway.

I'll leave this thread open for others to weigh in on.

[the remark this answers is not in the document]
Comment (katcharov) [reply driver-core/src/main/com/mongodb/internal/async/SingleResultCallback.java:64]: Fair enough, this sounds like a good idea for cases where we plan to use this outside of the API. Added.

File: driver-core/src/main/com/mongodb/internal/async/AsyncRunnable.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+package com.mongodb.internal.async;
+
+import com.mongodb.internal.async.function.RetryState;
+import com.mongodb.internal.async.function.RetryingAsyncCallbackSupplier;
+
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+/**
+ * See tests for usage (AsyncFunctionsTest).
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/async/AsyncRunnable.java:26]: I prefer that usage documentation is placed in the code rather than the test.  It will be easier for maintainers to refer to it if it's here.  And that's where all of our existing internal Javadoc is, so it will be less surprising.

Of all the new interfaces, this seems like the best place to put it.

**Review by jyemin — CHANGES_REQUESTED** [review]
Resolve one thread and commented back on another.

File: driver-core/src/main/com/mongodb/internal/async/SingleResultCallback.java [hunk 4]
```diff
@@ -34,4 +36,40 @@ public interface SingleResultCallback<T> {
...
+     * @return this callback as a handler
+     */
+    default AsyncCompletionHandler<T> asHandler() {
+        return new AsyncCompletionHandler<T>() {
+            @Override
+            public void completed(@Nullable final T result) {
+                if (result != null) {
+                    complete(result);
+                } else {
+                    complete((SingleResultCallback<Void>) SingleResultCallback.this);
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/async/SingleResultCallback.java:50]: Replacing the whole block with 

```
onResult(result, null)
```

seems preferable to this cast.

**Review by jyemin — CHANGES_REQUESTED** [review]

**Comment by rozza** [comment]
Firstly, I really like where this API is going.

I initially got confused as the language wasn't as I was expected coming from a functional background (eg. Scala and RxJava / Reactor). This got me thinking, and the API could be simplified by adopting some methods akin to those in [Project Reactors Mono](https://projectreactor.io/docs/core/release/api/) eg `map`, `flatmap` and `onErrorResume`.

I did a quick [POC](https://github.com/rozza/mongo-java-driver/blob/AsyncAPIMono/driver-core/src/main/com/mongodb/internal/async/Mono.java) based off this work and it provides the following methods:

| Method | Description |
| -----| ---|
| `Mono<Void> empty()` | produces a new `Mono` and starts the chain |
| `void finish(final SingleResultCallback<T> callback)` | Use the Mono with the supplied callback |
| `Mono<R> map(final Function<T, R> function)` | synchronously converts value `T` to `R`. |
| `Mono<R> flatMap(Function<T, Mono<R>> transformer)` | takes value `T` and provides a `Mono<R>`. |
| `Mono<T> then(final Mono<T> mono)`| a helper method that uses `flatMap` ignores the previous result and provides a new `Mono<T>`.|
| `Mono<T> onErrorResume(Function<? super Throwable, Mono<T>> fallbackFunction)` | handles any errors and produces a new `Mono<T>`. |
| `Mono<T> doWhile(final Mono<T> mono, final Predicate<Throwable> shouldRetry)` | handles any looping. |

The mapping between this API would be:

| AsyncRunnable | Mono |
| ------------- | ------------- |
| beginAsync                | empty   |
| unsafeFinish              | unsafeFinish |
| finish                    | finish |
|||
| getAsync                  | n/a |
| onErrorIf                 | onErrorResume |
| thenAlwaysRunAndFinish    | map().finish() / flatMap().finish() |
| thenRunAndFinish          | map().finish() / flatMap().finish() |
| thenApply                 | map / flatMap |
| thenConsume               | map / flatMap |
| thenRun                   | map / flatMap |
| thenRunIf                 | map / flatMap |
| thenSupply                | map / flatMap |
| thenRunRetryingWhile      | doWhile |
    

You can compare the tests here: [AsyncFunctionsTest](https://github.com/mongodb/mongo-java-driver/blob/JAVA-5082-2/driver-core/src/test/unit/com/mongodb/internal/async/AsyncFunctionsTest.java) and [MonoTest](https://github.com/rozza/mongo-java-driver/blob/AsyncAPIMono/driver-core/src/test/unit/com/mongodb/internal/async/MonoTest.java)

Basically, its very similar but follows a more conventional reactive API rather that trying to coherse sync Functional types (Runnable, Supplier, Consumer) into an async paradigm.

Let me know what you think?

File: driver-core/src/main/com/mongodb/internal/async/AsyncRunnable.java (hunk 3 above)
Comment (katcharov) [reply driver-core/src/main/com/mongodb/internal/async/AsyncRunnable.java]: Moved; I find that I often look at the test code (more often than the docs) to confirm that I am using the API correctly.

**Comment by katcharov** [comment]
@rozza Interesting! Thanks for digging into the options, this is the sort of thing that will help us round out the API. The most important thing, in my view, is the function composition that happens in "then" methods, which allows us to flatten nested async calls, and monadically short-circuit ensuing calls (which is comparable to exception handling). Both API surfaces build on this, so ultimately the APIs are much closer to each other than, for example, our current approach, or using reactive. Of course there are specific differences that arise, which can be individually addressed:

1. I think we should avoid the "indented" formatting approach. Consider:

```
try {
    plain(0);
    sync(1);
} catch (Throwable t) {
    sync(2);
}
sync(3);

// ========== Option 1:

Mono.empty().then(c2 -> {
    Mono.empty().then(c -> {
        plain(0);
        async(1, c);
    }).onErrorResume(t -> (c) -> {
        async(2, c);
    }).finish(c2);
}).then(c -> {
    async(3, c);
}).finish(callback);

// ========== Option 2:

Mono.empty()
        .then(c2 ->
                Mono.empty()
                        .then(c -> {
                            plain(0);
                            async(1, c);
                        })
                        .onErrorResume(t -> (c) -> async(2, c))
                        .finish(c2))
        .then(c -> async(3, c))
        .finish(callback);
```

While Option 2 can be shorter (though typically by only 1 line) in simple cases, in typical non-simple cases, where clarity is all the more important, it takes up more lines, and has substantial indentation. It is also less consistent, since it partly uses the "Option 1" approach when there is plain code mixed in (see then, vs onErrorResume above), which often happens. The biggest down-side is that it mixes boilerplate with normal code, which must be compared against a sync counterpart, whereas Option 1 puts all boilerplate on its own lines. I find matching the first option up with its async counterpart trivial (I just skim vertically along the suitable indentation level), but the second much more difficult.

2. The naming and conceptualization is similar to the sync functional types (like Runnable) in part because there are cases where we use a `Runnable` in sync code, but in the async code, the Runnable must return via callback, which is actually just the AsyncRunnable interface here (and see also the existing `AsyncCallbackRunnable`). It seems awkward or incorrect to replace it with a Mono. It also seems confusing to think of this as a Mono, which is supposed to be "Publisher with basic rx operators that emits at most one item via the onNext signal then terminates with an onComplete signal". I think we should avoid reactive concepts like publishers and signals, and avoid even certain functional concepts don't seem to fit. For example, I somewhat understand the general idea of using flatmap for thenSupply, but conceptually I find it all much more confusing - functionally, what are we mapping over when an input is "void", perhaps a side-effect?  In my head, only "thenApply" corresponds cleanly to an (asynchronous) map.

3. Some of the other changes I don't have a clear position on. For example, I think a `thenIf` helper makes it easier to match sync code, since it avoids having a extraneous `else` with completion boilerplate, but maybe this doesn't happen so often that it really needs a helper. These might be easier to consider in context, as review comments on relevant code.

File: driver-core/src/main/com/mongodb/internal/async/AsyncSupplier.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+    /**
+     * @param errorCheck A check, comparable to a catch-if/otherwise-rethrow
+     * @param errorFunction   The branch to execute if the error matches
+     * @return The composition of this, and the conditional branch
+     */
+    default AsyncSupplier<T> onErrorIf(
+            final Predicate<Throwable> errorCheck,
+            final AsyncFunction<Throwable, T> errorFunction) {
+        return (callback) -> this.finish((r, e) -> {
```
Comment (stIncMale) [inline driver-core/src/main/com/mongodb/internal/async/AsyncSupplier.java:116]: It's surprising that we need `finish` here instead of `unsafeFinish`, given that everywhere else `unsafeFinish` is, as expected, sufficient<sup>1</sup>. Could you leave a comment explaining why `finish` is required here?

---

<sup>1</sup> `AsyncRunnable.thenRunRetryingWhile` also calls `finish`, but there the situation is very different: we must implement `AsyncCallbackSupplier.get` there, and only `finish` meets the contract of `AsyncCallbackSupplier.get`.

**Review by stIncMale — CHANGES_REQUESTED** [review]

**Comment by rozza** [comment]
@katcharov, thanks for the feedback.My aim was not to create new language (and be barrier for learning / adoption) but rather to used established names functional / composable code. 

I picked the name `Mono` as its a common name that represents a type of `Future` which eventually emits a single result or errors (even for `Mono<Void>` produces a result (null) or an error). The functions `map` and `flatMap` are common across many languages and are applicable to both sync and async code. In Scala it is the foundation of composable Futures (the language even has syntactic sugur that underneath uses flatMap). 

The `Mono` interface doesn't add any reactive concepts like publishers or signals. It just uses `map` to transform actual values and `flatMap` to convert resulting values into another `Mono` composition and chaining the value supplied by the previous `Mono`. I added `then` to handle cases where you don't care about the previous result (eg `Mono<Void>` just produces `null`) but want to chain the logic - its just a `flatMap` where the input is ignored. The `Mono` class could be extended to handle predicates for mapping or error handling but the core methods required are already in its API.

Taking the example code - there are no runnables, just a flow of methods / and a catch:
```
try {
    plain(0);
    sync(1);
} catch (Throwable t) {
    sync(2);
}
sync(3);

// Can be flatted to its simplest form of (no need to add a nested Mono.empty()):
Mono.empty()
    .then(c -> {                        // try
        plain(0);
        async(1, c);
    }).onErrorResume(t -> (c) -> {      // catch
        async(2, c);
    })
    .then(c -> {                        // do after
        async(3, c);
    })
.finish(callback);
```

Even with `SingleResultCallback`s there are still cases where code is run synchronously (eg the `plain` method in the tests) and there are cases where the results are asynchronously produced (eg.  the `async` method). The example is a little artifical as previous side effects of the methods are ignored, so it lacks the passing of values from callback to callback, which is common in the codebase. 

Similar to `AsyncRunnable`, `Mono` allows for easy replacement of the `SingleResultCallback` API in stages. However, a benefit by using a single type would be to eventually just pass the `Mono`'s around to build up the composition of asynchronous actions and not have to switch in and out of callback based code.

File: driver-core/src/main/com/mongodb/internal/async/SingleResultCallback.java [hunk 6]
```diff
@@ -34,4 +36,36 @@ public interface SingleResultCallback<T> {
...
+
+    default void complete(final SingleResultCallback<Void> callback) {
+        // takes a void callback (itself) to help ensure that this method
+        // is not accidentally used when "complete(T)" should have been used
+        // instead, since results are not marked nullable.
+        Assertions.assertTrue(callback == this);
+        this.onResult(null, null);
+    }
+
+    default void complete(final T result) {
```
Comment (stIncMale) [inline driver-core/src/main/com/mongodb/internal/async/SingleResultCallback.java:64]:
```suggestion
    default void complete(@Nullable final T result) {
```

File: driver-core/src/main/com/mongodb/internal/async/AsyncSupplier.java [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+            if (e == null) {
+                callback.complete(r);
```
[the remark this answers is not in the document]
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/internal/async/AsyncSupplier.java:118]: `complete` actually allows `null` arguments.

File: driver-core/src/main/com/mongodb/internal/async/AsyncSupplier.java [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+    /**
+     * This is the async variant of a supplier's get method.
+     * This method must only be used when this AsyncSupplier corresponds
+     * to a {@link java.util.function.Supplier} (and is therefore being
+     * used within an async chain method lambda).
+     * @param callback the callback
+     */
+    default void getAsync(final SingleResultCallback<T> callback) {
+        unsafeFinish(callback);
+    }
```
[the remark this answers is not in the document]
Comment (katcharov) [reply driver-core/src/main/com/mongodb/internal/async/AsyncSupplier.java:47]: I see no reason not to.

File: driver-core/src/test/functional/com/mongodb/client/TestListener.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * A simple listener that consumes string events, which can be checked in tests.
+ */
+@ThreadSafe
+public final class TestListener {
```
[the remark this answers is not in the document]
Comment (katcharov) [reply driver-core/src/test/functional/com/mongodb/client/TestListener.java:29]: I also make use of it in [OIDC](https://github.com/mongodb/mongo-java-driver/pull/1134/files#diff-29fd9d33e419e53cfa212ebc47273c7c2b5e94f2ddf3dbdb197030c940a7edc1R871).

**Review by stIncMale — APPROVED** [review]

**Comment by rozza** [comment]
I like the general approach and had some naming concerns that are pretty trivial.

I will add a ticket for future consideration regarding adapting this approach, and extending it to provide a chainable (Monadic) API. That does not block this work, and as there are 2 LGTMs I've removed myself from the review.
